pkg/resource/v1/certconfig/desired.go

```diff
@@ -2,8 +2,347 @@ package certconfig
+	systemMastersOrganization = "system:masters"
+)
+
+var (
+	kubeAltNames = []string{
+		"kubernetes",
+		"kubernetes.default",
+		"kubernetes.default.svc",
+		"kubernetes.default.svc.cluster.local",
+	}
 )
 
+// GetDesiredState returns all desired CertConfigs for managed certificates.
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig, err := prepareClusterConfig(*r.baseClusterConfig, *clusterGuestConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredCertConfigs := make([]*v1alpha1.CertConfig, 0, len(managedCertificates))
+	for _, mc := range managedCertificates {
+		certConfig := mc.certConfigFactoryFunc(clusterConfig, mc.name, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+
+	return desiredCertConfigs, nil
+}
+
+func prepareClusterConfig(baseClusterConfig cluster.Config, clusterGuestConfig v1alpha1.ClusterGuestConfig) (*cluster.Config, error) {
+	var err error
+
+	// Copy baseClusterConfig as basis and supplement it with information from
+	// clusterGuestConfig.
+	clusterConfig := baseClusterConfig
+
+	clusterConfig.ClusterID = key.ClusterID(clusterGuestConfig)
+
+	clusterConfig.Domain.API, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.APICert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Calico, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.CalicoCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Etcd, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.EtcdCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.NodeOperator, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.NodeOperatorCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Prometheus, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.PrometheusCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.ServiceAccount, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.ServiceAccountCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Worker, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.WorkerCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	versionBundle, err := versionbundle.GetBundleByName(key.VersionBundles(clusterGuestConfig), certOperatorID)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig.VersionBundleVersion = versionBundle.Version
+
+	return &clusterConfig, nil
+}
+
+func newAPICertConfig(clusterConfig *cluster.Config, cert certs.Cert, projectName string) *v1alpha1.CertConfig {
```

Comment:
These constructor functions look very similar, but there are subtle differences in `v1alpha1.CertConfigSpecCert` fields for each certificate.

Reply:
This is fine. IMO the PR implements a lot of spaghetti code and it would be way simpler to read and understand if we repeat a bit more code and make it more explicit. Now there are lists with functions and we iterate and execute callbacks and go down a deeper code path but could instead just call the functions here do some error handling and be done. The code which is related to each other would then also be closer together and not spread across files when there is actually no need for this to be separated because we just call this once and do not need such things in different places.

Reply:
To make sure I understood correctly: You suggest to remove `managedCertificates` and just make explicit calls to these `new*` functions in `GetDesiredState()`? I'm fine with this. I think I had some reason for the callback list but cannot figure it out now so it probably isn't relevant anymore and can go.

Reply:
I think we are on the same page.